lte/cloud/go/sms_ll/sms_rp.go

```diff
@@ -0,0 +1,365 @@
+	OriginatorAddress RPAddressElement
+
+	// Mandantory. If Network->UE, must be length 0
+	DestinationAddress RPAddressElement
+
+	// Mandantory for RP-DATA, includes TPDU. If RP-ACK or RP-ERROR, must
+	// include IEI.
+	UserData RPUserElement
+
+	// Mandantory for RP-ERROR
+	Cause RPCauseElement
+}
+
+func (rpm RPMessage) MarshalBinary() ([]byte, error) {
+	b := []byte{rpm.MTI, rpm.Reference}
+
+	rpmt, err := rpm.MsgType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch rpmt {
+	case RP_DATA:
+		oa_data, err := rpm.OriginatorAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		da_data, err := rpm.DestinationAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		b = append(b, oa_data...)
+		b = append(b, da_data...)
+
+		ud_data, _ := rpm.UserData.MarshalBinary(RP_DATA)
+		b = append(b, ud_data...)
+	case RP_ERROR:
+		cause_data := rpm.Cause.MarshalBinary()
+		b = append(b, cause_data...)
+	case RP_ACK:
+		// Do nothing
+	default:
+		return nil, SMSRPError("Invalid RP Message Type")
+	}
+
+	return b, nil
+
+}
+
+func (rpm *RPMessage) UnmarshalBinary(input []byte) error {
+	if len(input) < 2 {
+		return SMSRPError("SMS-RP Message too short")
+	}
+
+	idx := 0
+	rpm.MTI = input[idx]
+	idx++ // 1
+	rpm.Reference = input[idx]
+	idx++ // 2
+
+	rpmt, err := rpm.MsgType()
+	if err != nil {
+		return err
+	}
+
+	switch rpmt {
+	case RP_DATA:
+		// The next two IEs should be adddresses in this case. So, get the lengths and pass to unmarshal
+		rpm.OriginatorAddress = *new(RPAddressElement)
+		rpm.DestinationAddress = *new(RPAddressElement)
+		rpm.UserData = *new(RPUserElement)
+		n, _ := rpm.OriginatorAddress.UnmarshalBinary(input[idx:])
+		if rpm.Direction() == RP_MO && n != 1 {
+			return SMSRPError("SMS-RP-DATA is MO, but OA length != 1")
+		}
+		idx += n
+		n, _ = rpm.DestinationAddress.UnmarshalBinary(input[idx:])
+		if rpm.Direction() == RP_MT && n != 1 {
+			return SMSRPError("SMS-RP-DATA is MT, but DA length != 1")
+		}
+		idx += n
+
+		n, _ = rpm.UserData.UnmarshalBinary(RP_DATA, input[idx:])
+		idx += n
+	case RP_ACK:
+		// RP-ACK and RP-ERROR may optionally contain an RP-User-Data
+		// element (TS24.001 7.3.3). If this is the case, it will be a
+		// TLV IE, with the first octet starting with the RP-User-Data
+		// IE ID (0x41).
+		if len(input) > 2 && input[idx] == RP_UDE_IEI {
+			rpm.UserData = *new(RPUserElement)
+			n, _ := rpm.UserData.UnmarshalBinary(RP_ACK, input[idx:])
+			idx += n
+		}
+	case RP_ERROR:
+		// Do nothing
+		rpm.Cause = *new(RPCauseElement)
+		idx += rpm.Cause.UnmarshalBinary(input[idx:])
+		// TODO: Add support for optional UserData TLV element.
+	default:
+		return SMSRPError(fmt.Sprintf("Invalid RP-SMS MTI: 0x%08b", rpm.MTI))
+	}
+
+	return nil
+}
+
+// If MTI is even, message is UE->Network
+func (rpm RPMessage) Direction() byte {
+	if rpm.MTI&0x1 == 0 {
+		return RP_MO
+	}
+	return RP_MT
+}
+
+func (rpm RPMessage) MsgType() (byte, error) {
```

Comment:
IMO you should save this into the RPMesage struct even though it's a derived value. Then you can trust your struct when you're serializing it. Do the validation during the unmarshal

Reply:
Unless you feel strongly about this, I prefer derived values (that are more of a convenience than anything) to live outside the struct. Validation is still done in `unserialize`, so the struct is valid, and we can just use these when/if they're needed.